Support returning the correct values for the different QuerySet methods when using .values() and .values_list().

syastrov opened the pull request.

syastrov committed:
Support returning the correct values for the different QuerySet methods when using .values() and .values_list().

syastrov committed:
Fix slicing on QuerySet. Fix django queries test, and remove some ignored errors that are no longer needed.

syastrov committed:
Remove accidental change in RawQuerySet.

syastrov committed:
Readded some still-necessary ignores to aggregation django test.

mkurnikov commented on django-stubs/db/models/query.pyi:
```diff
@@ -46,7 +48,7 @@ class FlatValuesListIterable(BaseIterable):
 _T = TypeVar("_T", bound=models.Model, covariant=True)
 
-class QuerySet(Iterable[_T], Sized):
+class QuerySet(Generic[_T, _Row], Sized):
```
Is it possible to preserve `QuerySet` being an `Iterable` here? Or even a `Collection`, as it implements all required methods.

syastrov replied:
Thanks for your comments. It looks like `Collection` will work, at least for the stubs tests. Let's see if Django tests pass.

mkurnikov replied:
Could you also leave `Sized` be, AFAIK `Collection` doesn't subclass it.

syastrov committed:
Add more tests of first/last/earliest/last/__getitem__, per mkurnikov's comments.

- Fix .iterator()

syastrov committed:
Re-add Iterator as base-class of QuerySet.

syastrov committed:
Make QuerySet a Collection.

syastrov committed:
- Fix return type for QuerySet.select_for_update().
- Use correct return type for QuerySet.dates() / QuerySet.datetimes().
- Use correct type params in return type for QuerySet.__and__ / QuerySet.__or__
- Re-add Sized as base class for QuerySet.
- Add test of .all() for all _Row types.
- Add test of .get() for all _Row types.
- Remove some redundant QuerySet method tests.

syastrov committed:
Merge branch 'master' into queryset_values_and_values_list

syastrov committed:
Automatically fill in second type parameter for QuerySet.

... if second parameter is omitted.

syastrov committed:
Merge branch 'master' into queryset_values_and_values_list
